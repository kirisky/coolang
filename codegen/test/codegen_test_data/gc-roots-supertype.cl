-- test roots in supertypes

class A {
 c1:C;
 foo:Int;
 c2:C;
 
 setC1( newC1 : C ) : Object
  {
    {
      c1 <- newC1;
    }
  };
};

class B inherits A {
 bar:Int;
};

class C  {
 a:A;
 setA( newA : A ) : Object
  {
    {
      a <- newA;
    }
  };
};

class D {
 foo:Int;
};

class E {
 foo:Int;
};

class Main inherits IO {
  main() : Object { {
    let
      b:B <- new B,
    in {
      let 
	    c:C <- new C,
		  in {
			c.setA(b);
			b.setC1(c);
		  };
	  out_string("C and B should live\n\n");
	  new E;
    };
	out_string("C and B should die now\n\n");
	new E;
  } };
};
