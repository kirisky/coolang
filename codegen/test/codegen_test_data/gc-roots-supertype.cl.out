Allocated an object of type: Main
Allocated an object of type: B
Allocated an object of type: C
C and B should live

Allocated an object of type: E
C and B should die now

Freeing:
obj
  is_reachable=0
  typename=E

Freeing:
obj
  is_reachable=0
  typename=C

Freeing:
obj
  is_reachable=0
  typename=B

Allocated an object of type: E
